### PR TITLE
Feature/1 delete files

### PR DESCRIPTION
**Pull request by ppxl**
This PR creates the basic functionality

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -24,7 +45,7 @@ node('docker') {
                 gitWithCredentials("fetch --all")
 
                 if (branch == "main") {
-                    echo "This branch has been detected as the master branch."
+                    echo "This branch has been detected as the main branch."
```
Comment (jsprey) [inline Jenkinsfile:48]: On line 52 you missed to change the parameter ` -Dsonar.branch.target=master` to `-Dsonar.branch.target=master`

File: docs/developing_de.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+**Testdateien mit bestimmten Zeitstempeln:**
+
+Testdateien mit bestimmten Zeitstempel können helfen, das Löschverhalten im echten Dateisystem darzustellen:
+
+```bash
+# Verzeichnis anlegen
+mkdir -p /tmp/conftemp
+# leere Dateien mit Zeitstempel 2021-03-03 03:03 anlegen
```
Comment (jsprey) [inline docs/developing_de.md:39]:
```suggestion
# Ins Verzeichnis wechseln
cd /tmp/conftemp

# leere Dateien mit Zeitstempel 2021-03-03 03:03 anlegen
```

File: docs/developing_en.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# develop confluence-temp-delete-job
+
+This document describes developmen-centric aspects.
```
Comment (jsprey) [inline docs/developing_en.md]:
```suggestion
This document describes development-centric aspects.
```

File: docs/developing_en.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+- Docker
+- Go compiler (currently 1.15.x)
+
+## Helpful scripts
+
+For testing on a real file system, these functions and settings can help.
+
+**Debug output:**
+
+With a log-level of `debug` single file and directory views become transparent. However, depending on the amount of files and directories looked at, an obscene amount of data is generated:
```
Comment (jsprey) [inline docs/developing_en.md]:
```suggestion
With a log-level of `debug` single file and directory views become transparent. However, depending on the number of files and directories looked at, an obscene amount of data is generated:
```

File: docs/developing_en.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+### Deletion in two phases
+
+The deletion routine of `tempdel` is essentially based on [`filepath.Walk`](https://golang.org/pkg/path/filepath/#Walk). In it, a file tree is walked recursively (and alphabetically, to act deterministically).
+
+The nature of Confluence temp files is still unclear. Therefore, a single delete operation consists of two phases:
+
+1. recursively delete files older than desired.
+1. recursively delete directories that are left empty.
+
+The two-phase approach has the advantage of deleting old directories that may contain new files. A complicating factor is that a file deletion updates the file stamp of a directory. Therefore the 1st phase is limited only to files. In the 2nd phase then exclusively empty directories are deleted, since on their timestamp no more reliance is anyway because of the file delete update.
```
Comment (jsprey) [inline docs/developing_en.md]: No entiendo :)

```suggestion
The two-phase approach has the advantage of deleting old directories that may contain new files. A complicating factor is that a file deletion updates the file stamp of a directory. Therefore the 1st phase is limited only to files. In the 2nd phase, only empty directories are deleted, since their timestamps can no longer be relied upon anyway due to the file delete update.
```

File: docs/developing_en.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+Translated with www.DeepL.com/Translator (free version)
```
Comment (jsprey) [inline docs/developing_en.md]: Achso :dagger:

File: docs/operations/delete-loop_en.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+This command accepts three basic input parameters:
+
+1. Start directory - the directory that should be searched for files
+1. File age - the maximum age in hours of the files that should not be deleted
+1. Deletion run interval - the time between the search intervals in minutes
+
+
+### Start directory
+
+The start directory is a mandatory parameter. The start directory can be freely selected and is not set to a specific value. There is no check if `tempdel` is executed against system relevant directories (**BETTER NOT,** because system damage may occur).
```
Comment (jsprey) [inline docs/operations/delete-loop_en.md]:
```suggestion
The start directory is a mandatory parameter. The start directory can be freely selected and is not set to a specific value. There is no check if `tempdel` is executed against system-relevant directories (**BETTER NOT,** because system damage may occur).
```

File: docs/operations/delete-loop_en.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+### File age
+
+The `--age`/`-a` switch can be used to optionally specify the maximum age (counted in hours from `now`) that files can have without being deleted. Only a positive integer value is accepted. Default value is `12` hours.
```
Comment (jsprey) [inline docs/operations/delete-loop_en.md]:
```suggestion
The `--age`/`-a` switch can be used to optionally specify the maximum age (counted in hours from `now`) that files can have without being deleted. Only a positive integer value is accepted. The default value is `12` hours.
```

File: docs/operations/delete-loop_en.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+### Deletion run interval
+
+The `--age`/`-a` switch can be used to optionally specify the interval (counted in minutes) between each deletion execution. Only a positive integer value is accepted. Default value is `60` minutes.
```
Comment (jsprey) [inline docs/operations/delete-loop_en.md]:
```suggestion
The `--age`/`-a` switch can be used to optionally specify the interval (counted in minutes) between each deletion execution. Only a positive integer value is accepted. The default value is `60` minutes.
```

File: app.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	`%{time:15:04:05.000} %{shortfunc} ▶ %{level:.4s} %{id:03x} %{message}`,
+)
+
+func configureLogging(c *cli.Context) error {
+	backend := logging.NewLogBackend(os.Stdout, "", 0)
+	backendFormatter := logging.NewBackendFormatter(backend, format)
+	logging.SetBackend(backendFormatter)
+	logLevel, err := logging.LogLevel(c.String("log-level"))
+	if err != nil {
+		fmt.Printf("%s: invalid log level specified, please use critical, error, warning, notice, info or debug", time.Now().Format(time.RFC3339))
```
Comment (jsprey) [inline app.go:64]: We would fit better with our conventions if you used the values for our root-log-level to configure the logging, i.e., DEBUG, ERROR, WARN, INFO

File: app.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+		return errors.Wrap(err, "failed to configure logging")
+	}
+	logging.SetLevel(logLevel, "")
+	return nil
+}
+
+// testing interface
+type exiter interface {
```
Comment (jsprey) [inline app.go:72]: Would it be better to define an interface where you need it? Should this interface move to the actual test file?

File: app_test.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+func Test_newExiter(t *testing.T) {
+	t.Run("should create an exiter instance", func(t *testing.T) {
+		sut := &defaultExiter{}
+
+		require.NotNil(t, sut)
+		assert.Implements(t, (*exiter)(nil), sut)
+	})
+}
+
+// test util stuff
```
Comment (jsprey) [inline app_test.go]: loose comment?

Comment (jsprey) [reply app_test.go]: Does it test util stuff or is it util stuff for the tests? I think the second. However, it does not help the understanding of the code. It should be removed or the intention should be expressed more clearly.

File: cmd/deleteFiles.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+const (
+	flagMaxAgeHoursLong   = "age"
+	flagMaxAgeHoursShort  = "a"
+	flagLoopIntervalLong  = "interval"
+	flagLoopIntervalShort = "i"
+)
+
+const cpuLoadSleepInSec = 2
```
Comment (jsprey) [inline cmd/deleteFiles.go:22]: Shouldn't it be moved to the const group? They have different intentions however they are all constants.

File: cmd/deleteFiles.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+			Aliases: []string{flagLoopIntervalShort},
+		},
+	},
+}
+
+func deleteFiles(c *cli.Context) error {
+	directory := ""
+	maxAgeInHours := c.Int(flagMaxAgeHoursLong)
+	loopIntervalInMin := c.Int(flagLoopIntervalLong)
+	loopIntervalInSecs := loopIntervalInMin * 60
```
Comment (jsprey) [inline cmd/deleteFiles.go]: Extract magic number :mage:

File: cmd/deleteFiles.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+		&cli.IntFlag{
+			Name:    flagMaxAgeHoursLong,
+			Usage:   "Sets the max. age of files and directories in hours that will be deleted. Must be larger than zero.",
+			Value:   12,
+			Aliases: []string{flagMaxAgeHoursShort},
+		},
+		&cli.IntFlag{
+			Name:    flagLoopIntervalLong,
+			Usage:   "Sets the interval in minutes to run the deletion routine. Must be larger than zero.",
+			Value:   60,
```
Comment (jsprey) [inline cmd/deleteFiles.go:45]: |
V

File: cmd/deleteFiles.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+		case <-ticker.C:
+			log.Debug("[tempdel] Start deletion run...")
+			err = deleteFilesWithArgs(args)
+			if err != nil {
+				log.Errorf("[tempdel] Deleting files failed with this error: %s", err.Error())
+			}
+			log.Debug("[tempdel] End deletion run.")
+		default:
+			// reduces CPU load but stretches reaction to unix signals
+			time.Sleep(cpuLoadSleepInSec * time.Second)
```
Comment (jsprey) [inline cmd/deleteFiles.go:119]: Nice, hack :)

**Review by jsprey — CHANGES_REQUESTED** [review]

File: deletion/deleter.go [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/hashicorp/go-multierror"
+	"github.com/op/go-logging"
+	errors2 "github.com/pkg/errors"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+var log = logging.MustGetLogger("deletion")
```
Comment (jsprey) [inline deletion/deleter.go]: Move this to the var group for consistency.

File: deletion/deleter_test.go [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	tests := []struct {
+		name        string
+		args        args
+		wantDeleter bool
+		wantErr     bool
+	}{
+		{"should pass", args{Args{Directory: "/test", MaxAgeInHours: 12}}, true, false},
+		{"should pass with 0 age", args{Args{Directory: "/test", MaxAgeInHours: 0}}, true, false},
+		{"should fail with dir", args{Args{Directory: "", MaxAgeInHours: 12}}, false, true},
```
Comment (jsprey) [inline deletion/deleter_test.go]:
```suggestion
		{"should fail with dir", args{Args{Directory: "", MaxAgeInHours: 12}}, false, true},
```
```suggestion
		{"should fail with invalid directory", args{Args{Directory: "", MaxAgeInHours: 12}}, false, true},
```

File: deletion/deleter_test.go [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+		{"should fail with age", args{Args{Directory: "/a", MaxAgeInHours: -1}}, false, true},
```
Comment (jsprey) [inline deletion/deleter_test.go]:
```suggestion
		{"should fail with invalid age", args{Args{Directory: "/a", MaxAgeInHours: -1}}, false, true},
```

File: deletion/deleter_test.go [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+		{"should fail with both dir and age", args{Args{Directory: "", MaxAgeInHours: -1}}, false, true},
```
Comment (jsprey) [inline deletion/deleter_test.go]:
```suggestion
		{"should fail with invalid directory and age", args{Args{Directory: "", MaxAgeInHours: -1}}, false, true},
```

**Review by jsprey — CHANGES_REQUESTED** [review]

**Comment by jsprey** [comment]
It was a pleasure to read your code. There are just some minor spelling things. I tested the temp-cleaner locally with success. Now I will start testing the cleaner together with confluence.

File: Jenkinsfile (hunk 1 above)
Comment (ppxl) [reply Jenkinsfile:48]: Sadly no. This is a reference to a SonarQube internal value which does not work with `main`. You can already see a SQ scan on the main branch here https://ecosystem.cloudogu.com/sonar/dashboard?branch=master&id=confluence-temp-delete-job

File: app.go (hunk 10 above)
Comment (ppxl) [reply app.go:64]: I see what you are trying to achieve, although I wholeheartedly disagree. 

The mentioned log levels are meant to be used on a Dogu level while the actually used log levels highly depend on the system being used and must be mapped individually. F. i. there are mappings from `root/logging/ERROR` to a Tomcat's `Severe`. Additionally all of the Cloudogu CLI applications embrace lower case values.

File: app.go (hunk 11 above)
Comment (ppxl) [reply app.go:72]: I created this interface exactly for the reason of introducing testability. An interface in a test file does not compile to the actual production code (which I need in order to avoid hard occurrences of `os.Exit(1)` which ends the test as well.

Anyhow I understand that the code comment could be worded a bit more carefully. I shall reword the doc comment.

File: app_test.go (hunk 12 above)
Comment (ppxl) [reply app_test.go]: point taken. I shall remove this comment

File: cmd/deleteFiles.go (hunk 13 above)
Comment (ppxl) [reply cmd/deleteFiles.go:22]: The way I see it grouping of declarations is a form of semantic association.

Reading the comment in this [discussion](https://github.com/golang/go/issues/5255#issuecomment-66077862) of go/doc I get the feeling that I am not too far from the intentions.

Grouping constants is neither necessary and does not introduce syntactical meaning either. Instead, grouping helps to make read the code better in order to see which things belong into the same semantic context and which does not. If we wanted we could produce many many constant groups if it served a meaningful purpose.

Long story short. This one const `cpuLoadSleepInSec` does not pay into the same thing as those earlier constants which refer to CLI parameter handling.

File: deletion/deleter.go (hunk 17 above)
Comment (ppxl) [reply deletion/deleter.go]: done.

File: docs/developing_de.md (hunk 2 above)
Comment (ppxl) [reply docs/developing_de.md:39]: done

File: .gitignore [hunk 21]
```diff
@@ -12,4 +12,8 @@
 *.out
 
 # Dependency directories (remove the comment below to include it)
-# vendor/
+vendor/
+
+target/
+.idea/
+.code/
```
Comment (ppxl) [inline .gitignore]: I also added here a newline

File: app.go (hunk 11 above)
Comment (ppxl) [reply app.go:72]: Done

File: app_test.go (hunk 12 above)
Comment (ppxl) [reply app_test.go]: done.

File: cmd/deleteFiles.go (hunk 15 above)
Comment (ppxl) [reply cmd/deleteFiles.go:45]: These values are embedded into a semantic context (which removes the magic from magic constants). I don't see them as magic constants. Also, they are nowhere used besides this code fragment.

If you want I can provide code like this but it does not introduce any more meaning than it does now:

```golang
const defaultLoopInterval60Minutes = 60;
flags := []cli.Flag{
	&cli.IntFlag{
		Name:    flagLoopIntervalLong,
		Usage:   "Sets the interval in minutes to run the deletion routine. Must be larger than zero.",
		Value:   defaultLoopInterval60Minutes,
		Aliases: []string{flagLoopIntervalShort},
	},
}
```

File: cmd/deleteFiles.go (hunk 14 above)
Comment (ppxl) [reply cmd/deleteFiles.go]: I rewrote the calculation. Please be so kind and check again.

File: docs/developing_en.md (hunk 6 above)
Comment (ppxl) [reply docs/developing_en.md]: LOL I remove id.

Comment (ppxl) [reply docs/developing_en.md]: done.

**Review by jsprey — APPROVED** [review]
